pkg/k8s/watchers/crd.go

```diff
@@ -0,0 +1,299 @@
+	// peroid.
+	ctx, cancel := context.WithTimeout(context.Background(), option.Config.CRDWaitTimeout)
+	for crd := range crds.m {
+		k.blockWaitGroupToSyncResources(
+			ctx.Done(),
+			nil,
+			func() bool {
+				crds.Lock()
+				defer crds.Unlock()
+				return crds.m[crd]
+			},
+			crd,
+		)
+	}
+
+	// The above loop will call blockWaitGroupToSyncResources to populate the
+	// K8sWatcher state with the current state of the CRDs. It will check the
+	// state of each CRD, with the inline function provided. If the function
+	// reports that the given CRD is true (has been synced), it will close a
+	// channel associated with the given CRD. A subsequent call to
+	// (*K8sWatcher).WaitForCacheSync will notice that a given CRD's channel
+	// has been closed. Once all the CRDs passed to WaitForCacheSync have had
+	// their channels closed, the function unblocks.
+	//
+	// Meanwhile, the Goroutine below kicks off the controller that was
+	// instantiated above, and enters a loop looking for (1) if the context has
+	// deadlined or (2) if the entire CRD state has been synced (all CRDs found
+	// in the cluster). While the Goroutine is in this loop, the controller is
+	// listening for either add or delete events to the
+	// customresourcedefinition resource (disguised inside a
+	// metav1.PartialObjectMetadata object).  If (1) is encountered, then
+	// Cilium will fatal because it cannot proceed if the CRDs are not present.
+	// If (2) is encountered, then make sure the controller has exited by
+	// cancelling the context and we return out.
+
+	go func() {
+		defer cancel()
+		go crdController.Run(ctx.Done())
+		k.k8sAPIGroups.addAPI(k8sAPIGroupCRD)
+
+		ticker := time.NewTicker(1 * time.Second)
+		for {
+			select {
+			case <-ctx.Done():
+				if err := ctx.Err(); err != nil {
+					log.WithError(err).
+						Fatalf("Unable to find all Cilium CRDs necessary within "+
+							"%v timeout. Please ensure that Cilium Operator is "+
+							"running, as it's responsible for registering all "+
+							"the Cilium CRDs.",
+							option.Config.CRDWaitTimeout)
+				}
+			case <-ticker.C:
+				if crds.isSynced() {
+					ticker.Stop()
+					log.Info("All Cilium CRDs have been found and are available")
+					return
+				}
+			}
+		}
+	}()
+}
+
+// isSynced returns whether all the CRDs inside `m` have all been synced,
+// meaning all CRDs we care about in Cilium exist in the cluster.
+func (s *crdState) isSynced() bool {
+	s.Lock()
+	defer s.Unlock()
+	for crd := range s.m {
+		if s.m[crd] == false {
+			return false
+		}
+	}
+	return true
+}
+
+// crdState contains the state of the CRDs inside the cluster.
+type crdState struct {
+	lock.Mutex
+
+	// m is a map which maps the CRD name to its synced state in the cluster.
+	// True means it exists, false means it doesn't exist.
+	m map[string]bool
+}
+
+func newCRDState(m map[string]bool) crdState {
+	return crdState{
+		m: m,
+	}
+}
+
+const (
+	cnpCRD  = "crd:" + v2.CNPName
```

Comment:
Create a helper function to join the `crd:` with the name and it can be used in [1]